### PR TITLE
[Feature] implement str_to_map()

### PR DESCRIPTION
**Pull request by fzhedu**
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:

Fixes #

## Problem Summary(Required):

 implement `MAP<VARCHAR, VARCHAR> str_to_map(VARCHAR content[, VARCHAR delimiter[, VARCHAR kv_delimiter]])`, more msg please refer to doc.
```
mysql> select str_to_map(c,',',b), c, b from t;
+------------------------------+---------------+------+
| str_to_map(split(c, ','), b) | c             | b    |
+------------------------------+---------------+------+
| NULL                         | a:b,null:null | NULL |
| {"":null}                    |               | ,    |
| {"a":"c:b:d","":null}        | a:c:b:d,,,    | :    |
| NULL                         | NULL          | :    |
| {"a":"","c:d过’":null}       | a:b,c:d过’    | :b   |
| {"a":":b"}                   | a:b           |      |
+------------------------------+---------------+------+
6 rows in set (0.02 sec)
```
besides solve some bugs to identify chinese characters in `split()` and `split_part()`.
```
mysql> select str_to_map('道c:d过',',','');
+-----------------------------------------+
| str_to_map(split('道c:d过', ','), '')   |
+-----------------------------------------+
| {"道":"c:d过"}                            |
+-----------------------------------------+
1 row in set (0.02 sec)
```
## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 16 (81.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/FunctionAnalyzer.java|1|4|25.00%|[105, 106, 107]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|12|12|100.00%|[]|

File: be/test/exprs/string_fn_test.cpp [hunk 1]
```diff
@@ -622,6 +622,126 @@ PARALLEL_TEST(VecStringFunctionsTest, splitConst2) {
     ASSERT_TRUE(StringFunctions::split_close(ctx.get(), FunctionContext::FunctionStateScope::FRAGMENT_LOCAL).ok());
 }
 
+TypeDescriptor array_type(const LogicalType& child_type);
```
Comment (imay) [inline be/test/exprs/string_fn_test.cpp:625]: How about writing test cases in sql-test that is more convenient?

Comment (fzhedu) [reply be/test/exprs/string_fn_test.cpp:625]: some cases should be written in UT, for example constructing const columns, only-null columns.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java [hunk 2]
```diff
@@ -99,6 +100,12 @@ public static void analyze(FunctionCallExpr functionCallExpr) {
                     "map_apply's lambda function can not be null");
             functionCallExpr.setType(functionCallExpr.getChild(0).getChild(0).getType());
         }
+
+        if (fnName.getFunction().equals(FunctionSet.STR_TO_MAP)) {
+            Type retType = new MapType(Type.VARCHAR, Type.VARCHAR);
+            functionCallExpr.setType(retType);
+            functionCallExpr.getFn().setRetType(retType);
+        }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java:108]: should resovle it in PolymorphicFunctionAnalyzer?

File: be/src/exprs/str_to_map.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    auto is_unique = [=](std::vector<Slice>& exist_slice, Slice s) {
+        for (auto tmp : exist_slice) {
+            if (s == tmp) {
+                return false;
+            }
+        }
+        exist_slice.push_back(s);
+        return true;
+    };
```
Comment (Seaven) [inline be/src/exprs/str_to_map.cpp:73]:
```suggestion
    auto is_unique = [=](std::vector<Slice>& exist_slice, Slice& s) {
        for (auto& tmp : exist_slice) {
            if (s == tmp) {
                return false;
            }
        }
        exist_slice.emplace_back(s);
        return true;
    };
```

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java (hunk 2 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java:108]: as it is speical that all its parameters are not Polymorphic.

File: be/src/exprs/str_to_map.cpp (hunk 3 above)
Comment (fzhedu) [reply be/src/exprs/str_to_map.cpp]: chagned

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java:108]: so, if this is a sepcial type function, why not register it by special type? like:
```
    [170004, 'str_to_map', 'MAP<VARCHAR, VARCHAR>', ['MAP<VARCHAR, VARCHAR>', 'VARCHAR', 'VARCHAR'], 'nullptr'],
```

be may we need support this way?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java (hunk 2 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/FunctionAnalyzer.java]: added map_varchar_varchar

File: be/src/exprs/str_to_map.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+#include "util/utf8.h"
+
+namespace starrocks {
+
+/**
+* @param: [array_string, delimiter]
+* @paramType: [ArrayBinaryColumn, BinaryColumn]
+* @return: MapColumn map<string,string>
+
+ the original str_to_map(str, del1, del2) is rewritten to str_to_map(split(str, del1), del2), the first input results
```
Comment (chaoyli) [inline be/src/exprs/str_to_map.cpp:34]: 1. What's this function used for?
2. Why need two delimiters?
3. Could the delimiters be mulitple characters?

**Comment by fzhedu** [comment]
> The result of `select str_to_map('道c:d过',',','');` seems wrong ?

it is right, please refer to its doc.

1. cast str to map, used in hive, databricks.
2. first del split maps, second del split key-value pair.
3. sure, all delimiters could be a string.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 19 (10.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|1|18|05.56%|[5315, 5316, 5317, 5318, 5319, 5320, 5321, 5322, 5323, 5324, 5325, 5326, 5327, 5328, 5330, 5332, 5333]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|1|1|100.00%|[]|

**Review by Seaven — APPROVED** [review]

**Review by silverbullet233 — APPROVED** [review]
LGTM

File: be/src/exprs/split_part.cpp [hunk 5]
```diff
@@ -65,7 +66,17 @@ StatusOr<ColumnPtr> StringFunctions::split_part(FunctionContext* context, const
             if (part_number > haystack.size) {
                 res.append_null();
             } else {
-                res.append(Slice(haystack.data + part_number - 1, 1));
+                int char_size = 0, h = 0;
+                for (auto num = 0; h < haystack.size && num < part_number - 1; h += char_size) {
+                    char_size = UTF8_BYTE_LENGTH_TABLE[static_cast<unsigned char>(haystack.data[h])];
```
Comment (mofeiatwork) [inline be/src/exprs/split_part.cpp:71]: does it affect performance ?

File: be/src/exprs/str_to_map.cpp (hunk 4 above)
Comment (mofeiatwork) [reply be/src/exprs/str_to_map.cpp:34]: what if extra space in the string, like `{1:2,      2:3}` ?

File: be/src/exprs/split_part.cpp (hunk 5 above)
Comment (fzhedu) [reply be/src/exprs/split_part.cpp:71]: It has to detect character size in O(n), besides it is a rare case where the delimiter is an empty string.

File: be/src/exprs/str_to_map.cpp (hunk 4 above)
Comment (fzhedu) [reply be/src/exprs/str_to_map.cpp:34]: more msg please refer to doc.
the extra space will be in map, as they are token into string.

**Review by murphyatwork — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**